biothings/web/analytics/channels.py

```diff
@@ -1,83 +1,99 @@
-        # A maximum of 20 hits can be specified per request.
-        for i in range(0, len(events), 20):
-            yield HTTPRequest(
-                "http://www.google-analytics.com/batch",
-                method="POST",
-                body="\n".join(events[i : i + 20]),  # noqa: E203
-            )
+        async with aiohttp.ClientSession() as session:
+            for i in range(0, len(events), 20):
+                data = "\n".join(events[i:i + 20])
+                url = "http://www.google-analytics.com/batch"
+                await self.send_request(session, url, data)
+
+    async def send_request(self, session, url, data):
+        async with session.post(url, data=data) as _:
+            pass
 
 
 class GA4Channel(Channel):
     def __init__(self, measurement_id, api_secret, uid_version=1):
         self.measurement_id = measurement_id
         self.api_secret = api_secret
         self.uid_version = uid_version
+        self.max_retries = 1
 
-    def handles(self, event):
+    async def handles(self, event):
         return isinstance(event, Event)
 
-    def send(self, payload):
-        """
-
-        Limitations:
-        https://developers.google.com/analytics/devguides/collection/protocol/ga4/sending-events?client_type=gtag
-        """
+    async def send(self, payload):
         events = payload.to_GA4_payload(self.measurement_id, self.uid_version)
-        # #batch-limitations section of the URL above
-        # A maximum of 25 hits can be specified per request.
         url = f"https://www.google-analytics.com/mp/collect?measurement_id={self.measurement_id}&api_secret={self.api_secret}"
-        for i in range(0, len(events), 25):
-            data = {
-                "client_id": str(payload._cid(self.uid_version)),
-                "user_id": str(payload._cid(1)),
-                "events": events[i : i + 25],  # noqa: E203
-            }
-            yield HTTPRequest(url, method="POST", body=orjson.dumps(data))
+        async with aiohttp.ClientSession() as session:
+            # The pagination of 25 is defined according to the context of the current application
+            # Usually, each client request is going to make just 1 request to the GA4 API.
+            # However, it's possible to collect data to GA4 in other parts of the application.
+            for i in range(0, len(events), 25):
+                data = {
+                    "client_id": str(payload._cid(self.uid_version)),
+                    "user_id": str(payload._cid(1)),
+                    "events": events[i:i + 25],
+                }
+                await self.send_request(session, url, orjson.dumps(data))
+
+    async def send_request(self, session, url, data):
+        retries = 0
+        base_delay = 1  # Base delay in seconds
+        while retries <= self.max_retries:
+            async with session.post(url, data=data) as response:
+                if response.status == 502:  # HTTP 502 - Bad Gateway
```

Comment:
Do we want to extend this beyond just 502 error codes?

Reply:
> Do we want to extend this beyond just 502 error codes?

Make sense